### PR TITLE
Limit recursion of raw_call

### PR DESCRIPTION
**Pull request by mina86**
In certain situations a misbehaving contract could lead to infinite recursion.  This is possible because `executor::allocate` function calls into contract’s code.  If a malicious contract calls a host function which calls `executor::allocate`, it will cause an infinite recursion and eventual crash.

To fix the issue, limit recursion depth to just two passes from host code to contract code.  This is the approach taken by CosmWasm.  See https://github.com/JumpCrypto/security-research/blob/master/advisories/2023-003-cosmwasm.md

### CODE REVIEW

File: vm-wasmi/src/vm.rs [hunk 1]
```diff
@@ -505,7 +505,22 @@ where
             .and_then(Extern::into_func)
             .ok_or(WasmiVMError::FunctionNotFound)?;
 
-        export.call(self.0.as_context_mut(), &params, &mut result)?;
+        // Avoid DOS caused by infinite recursion.  See
+        // https://github.com/JumpCrypto/security-research/blob/master/advisories/2023-003-cosmwasm.md?ref=jumpcrypto.com
+        let mut ctx = self.0.as_context_mut();
+        let depth = ctx.data_mut().call_depth_mut();
+        if *depth >= 2 {
```
Comment (blasrodri) [inline vm-wasmi/src/vm.rs]: maybe we can make this configurable? I'm just wondering if there could be a reason for this to recurse on a good case?

File: vm-wasmi/src/vm.rs [hunk 2]
```diff
@@ -505,7 +505,22 @@ where
...
+            return Err(WasmiVMError::MaxCallDepthExceeded.into());
+        }
+        *depth += 1;
+
+        let res = export.call(ctx, &params, &mut result);
+
+        let mut ctx = self.0.as_context_mut();
+        let depth = ctx.data_mut().call_depth_mut();
+        *depth = depth.checked_sub(1).unwrap();
```
Comment (blasrodri) [inline vm-wasmi/src/vm.rs]: why `checked_sub` if unwrap is used?

Comment (mina86) [reply vm-wasmi/src/vm.rs]: In release build `0u32 - 1` is `u32::MAX`.  I wanted to make sure we detect if there’s ever bug here.

Comment (blasrodri) [reply vm-wasmi/src/vm.rs]: Perfect. Do you want to ship this `.unwrap()` to prod? Or would you remove it once tests are added?

Comment (mina86) [reply vm-wasmi/src/vm.rs]: I was planning to keep it.  Don’t have strong feelings either way.

Comment (blasrodri) [reply vm-wasmi/src/vm.rs]: So, let's assume that the unwrap gets triggered, what happens @dzmitry-lahoda on the runtime?

File: vm-wasmi/src/vm.rs (hunk 1 above)
Comment (mina86) [reply vm-wasmi/src/vm.rs]: If you mean configured at run-time than in principle yes but I just don’t know where that option should be specified.

File: vm-wasmi/src/vm.rs (hunk 2 above)
Comment (mina86) [reply vm-wasmi/src/vm.rs]: The runtime crashes but if the unwrap gets triggered than it indicates bug in the code so controlled crash is better than unexpected buggy behaviour later.  Keep in mind this unwrap is unrelated to bug with infinite recursion; this isn’t something malicious contract can trigger.

**Review by blasrodri — APPROVED** [review]

File: vm-wasmi/src/vm.rs (hunk 2 above)
Comment (dzmitry-lahoda) [reply vm-wasmi/src/vm.rs]: better use exect and ensure you are not panic https://substrate.stackexchange.com/questions/317/when-to-use-expect

File: vm-wasmi/src/vm.rs (hunk 1 above)
Comment (dzmitry-lahoda) [reply vm-wasmi/src/vm.rs]: i am ok with constant. error case is here. clear when  to change. cosmos people also have constant.

**Comment by mina86** [comment]
PTAL, took me a while but eventually managed to write the test.
Without the fix test crashes:

    thread 'semantic::test_recursion' has overflowed its stack
    fatal runtime error: stack overflow

with the fix the call gracefully fails with SimpleVMError::Interpreter
error.

File: vm-wasmi/src/vm.rs (hunk 2 above)
Comment (mina86) [reply vm-wasmi/src/vm.rs]: Honestly I’m unconvinced that `expect` provides any value.  Failing `unwrap` will include file and line which will be enough to identify where the failure happened.  But sure, can change it to `expect`.

**Comment by blasrodri** [comment]
Amazing! Feel free to merge it :)

**Pull request merged**